wasm/src/language_service.rs

```diff
@@ -59,6 +62,27 @@ impl LanguageService {
         self.0.close_document(uri);
     }
 
+    pub fn update_notebook_document(&mut self, notebook_uri: &str, cells: Vec<ICell>) {
+        let cells: Vec<Cell> = cells.into_iter().map(|c| c.into()).collect();
```

Comment:
We are double materializing (collect) the cells here. Does it make sense to continue mapping so that we only iterate once and pass the result into the update call?

Reply:
Yeah I'm glad you brought it up, because this one drove me crazy. But I don't think it's possible to make this work without double-materializing. If I try to chain the `map`s like below, then the `.as_ref()` calls create a problem as we end up with a reference to a temporary value (created by `Cell::from(c)`). In order for the strings in `Cell` _not_ to be temporary, I need to materialize once after I convert from `JsValue` to `Cell`. Then, I need to materialize again since the callee `update_notebook_document` expects a slice, not an iterator.

```rs
let cells = cells
            .into_iter()
            .map(|c| Cell::from(c))
            .map(|s| (s.uri.as_ref(), s.version, s.code.as_ref()))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return value referencing local data
                                      `s.uri` returns a value referencing data owned by the current function
            .collect::<Vec<(&str, u32, &str)>>();
```

Reply:
I think I'll have the inner `update_notebook_document` take an iterator instead of a slice. That should fix it.